Greatly reduce mallocs in SQLString.canonicalizeString()

clockfort opened the pull request:
partially addresses #735 , which identifies this as a hotspot for heap allocations

This new impl Is ~6.5x fewer allocs than the previous impl (and because it was memory-bound problem, is now ~6.5x faster), and rather than being bug-for-bug compat, adds some functionality. (cleans up multiple UNREGISTERED_SQL_COMMENT in large batches of statements now rather than just the first).

I'm not in love with removing the trailing semicolon; It really messes with the query's feng shui. So I'd be more than open to removing that piece of functionality.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/1560)
<!-- Reviewable:end -->

SerialVelocity commented:
Hey @clockfort, would you be able to benchmark this impl:
```java
    public static String canonicalizeString(String original) {
        StringBuilder cleanedString = new StringBuilder(original);

        int originalIdx = 0;
        int cleanedIdx = 0;
        int firstUnregisteredIndex = cleanedString.indexOf(UNREGISTERED_SQL_COMMENT);
        while (originalIdx < original.length()) {
            char originalChar = original.charAt(originalIdx);
            if (originalIdx == firstUnregisteredIndex) {
                originalIdx += UNREGISTERED_SQL_COMMENT.length();
                firstUnregisteredIndex = original.indexOf(UNREGISTERED_SQL_COMMENT, originalIdx);
            } else if (originalChar == ' ' && (cleanedIdx == 0 || cleanedString.charAt(cleanedIdx - 1) == ' ')) {
                ++originalIdx;
            } else {
                cleanedString.setCharAt(cleanedIdx, originalChar);
                ++cleanedIdx;
                ++originalIdx;
            }
        }

        if (cleanedString.charAt(cleanedIdx - 1) == ' ') {
            --cleanedIdx;
        }

        if (cleanedString.charAt(cleanedIdx - 1) == ';') {
            --cleanedIdx;
        }

        return cleanedString.substring(0, cleanedIdx);
    }
```

From simple testing it seems to be 25% faster by combining everything into a single loop and not doing deletions.

(when I tested your implementation, I also changed `cleanedString.indexOf(UNREGISTERED_SQL_COMMENT)` to `cleanedString.indexOf(UNREGISTERED_SQL_COMMENT, prefixIndex)` to make it a bit faster)

clockfort commented:
On my machine yours is 2% faster for the normal case of non-batched single SQL statements and 17% faster for batches of two statements. (and would be more of a speedup for bigger batches)

- made your variable names consistent
- added an extra test case that your code made me think of, because your code made the problem more obvious. (both of our methods failed the test)
- added an extra if statement to fix the code for the test

schlosna commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -251,22 +252,40 @@ private static String makeCommentString(String keyString, DBType dbType) {
-        String trim = s.trim();
-        if(!trim.isEmpty() && trim.charAt(trim.length() - 1) == ';') {
-            trim = trim.substring(0, trim.length() - 1);
+    public static String canonicalizeString(String original) {
```
looks like the only caller of this method in AtlasDB immediately does a copy of the returned String and strips out all whitespace, so might be worthwhile to create a private version that does not copy over any whitespace. Note that there is at least one caller of `canonicalizeString` outside of this project so we'd still need the public method to return a String

```
        FinalSQLString cached = cachedUnregistered.get(StringUtils.deleteWhitespace(canonicalizeString(sql)));
```

SerialVelocity commented:
If you're trying to deal with more cases, then these exist:
```
a; b vs a;b
a;;b
ab;;
ab; ;
;a
```

clockfort commented:
probably not worth our time to decide on how to deal with edge cases that can only occur in batches stuff; I believe almost all usage for this is single statements.

- updated with a version that provides full whitespace removal and updated the unregistered query cache to use it.
- rebased (and fix incorrect automatically 3-way merged release notes)

schlosna commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -200,7 +201,7 @@ public static boolean isValidKey(final String key) {
     static FinalSQLString getUnregisteredQuery(String sql) {
         assert !isValidKey(sql) : "Unregistered Queries should not look like keys"; //$NON-NLS-1$
-        FinalSQLString cached = cachedUnregistered.get(StringUtils.deleteWhitespace(canonicalizeString(sql)));
+        FinalSQLString cached = cachedUnregistered.get(canonicalizeString(sql));
```
I think you want to call canonicalizeStringAndRemoveWhitespaceEntirely here

schlosna replied:
also need to remove unused `import org.apache.commons.lang3.StringUtils;`

jeremyk-91 commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -251,22 +252,49 @@ private static String makeCommentString(String keyString, DBType dbType) {
+        return canonicalizeString(sql, false);
+    }
+
+    static String canonicalizeStringAndRemoveWhitespaceEntirely(String sql) {
```
nit: we should add `@VisibleForTesting` here.

jeremyk-91 commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -251,22 +252,49 @@ private static String makeCommentString(String keyString, DBType dbType) {
+                originalIdx += UNREGISTERED_SQL_COMMENT.length();
+                firstUnregisteredIdx = original.indexOf(UNREGISTERED_SQL_COMMENT, originalIdx);
+            } else if (originalChar == ' ' && (cleanedIdx == 0 || cleanedString.charAt(cleanedIdx - 1) == ' ')
+                    || (originalChar == ' ' && removeAllWhitespaceEntirely)) {
```
The above looks like `(A and (B or C) or (A and D))`; is there a reason we didn't write it as `(A and (B or C or D))`?

jeremyk-91 commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -251,22 +252,49 @@ private static String makeCommentString(String keyString, DBType dbType) {
+        StringBuilder cleanedString = new StringBuilder(original);
+        int originalIdx = 0;
+        int cleanedIdx = 0;
+        int firstUnregisteredIdx = cleanedString.indexOf(UNREGISTERED_SQL_COMMENT);
```
nit: This name becomes less accurate if you have multiple instances of `UNREGISTERED_SQL_COMMENT`; I can see that it is the first in the remainder of the string, though. I think I'd prefer `nextUnregisteredIdx` (of course not a major issue).

jeremyk-91 commented on commons-db/src/test/java/com/palantir/nexus/db/sql/SQLStringTest.java:
```diff
@@ -0,0 +1,60 @@
+        List<String> testBatch = ImmutableList.of(
+                "/* UnregisteredSQLString */ insert foo into bar; /* UnregisteredSQLString */insert foo into bar;",
+                "insert foo into bar; /* UnregisteredSQLString */ insert foo into bar");
+        String canconicalBatch = "insert foo into bar; insert foo into bar";
```
nit: `canonical`

jeremyk-91 commented on commons-db/src/test/java/com/palantir/nexus/db/sql/SQLStringTest.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Test
+    public void testCanonicalizeStringNoSpaces() {
```
nit: seeing a test with this name I'd guess it be an analogue of the one-query test cases with removeWhitespaceEntirely rather than the batch test.

jeremyk-91 commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -42,8 +42,9 @@ develop
-         -
+    *    - |improved|
+         - Improved heap usage during heavy DBKVS querying
+           (Pull Request <https://github.com/palantir/atlasdb/pull/1560>`__)
```
Would be nice if we could quantify this, though I'd be happy to get this in anyway - this looks like it should address the flag in #735 and we can add the magnitude of the improvement before we next release

SerialVelocity commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -281,13 +282,13 @@ private static String canonicalizeString(String original, boolean removeAllWhite
         }
 
-        if (cleanedString.charAt(cleanedIdx - 1) == ' ') {
+        while (cleanedIdx > 0 && Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1))) {
```
This one only needs to be an `if` since all extra whitespace will be compressed down to one above

jeremyk-91 commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -210,75 +219,104 @@ static FinalSQLString getUnregisteredQuery(String sql) {
-        if (trim.startsWith(prefix)) {
-            trim = trim.substring(prefix.length()).trim();
+
+        while (cleanedIdx > 0 && Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1))) {
```
As @SerialVelocity pointed out, this can be an `if` since the checks on lines 283-284 would maintain that the cleaned part of the string can't have consecutive whitespace.

schlosna commented on commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java:
```diff
@@ -291,7 +291,7 @@ private static String canonicalizeString(String original, boolean removeAllWhite
         }
 
-        while (cleanedIdx > 0 && Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1))) {
+        if (cleanedIdx > 0 && Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1))) {
```
nit: really lines 294 - 303 could all be replaced with:

```
while (cleanedIdx > 0
    && (Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1))
            || cleanedString.charAt(cleanedIdx - 1) == ';')) {
    --cleanedIdx;
}
```
